File: javalin/src/main/java/org/example/joke/service/JokeService.java
package org.example.joke.service;

import org.example.joke.model.Joke;
import org.example.joke.model.JokeCreate;
import org.example.joke.repo.JokeRepository;

import java.util.List;
import java.util.Optional;

public class JokeService {
    private final JokeRepository repo;

    public JokeService(JokeRepository repo) {
        this.repo = repo;
    }

    public Joke create(JokeCreate create) {
        Joke j = Joke.of(create.setup(), create.punchline(), create.category(), create.source());
        return repo.create(j);
    }

    public Optional<Joke> get(String id) { return repo.findById(id); }

    public List<Joke> list(Optional<String> source, Optional<String> category, int limit, int offset) {
        return repo.list(source, category, limit, offset);
    }

    public Optional<Joke> update(String id, JokeCreate create) {
        return repo.findById(id).flatMap(old -> {
            Joke updated = new Joke(id, create.setup(), create.punchline(), create.category() == null ? "" : create.category(), create.source() == null ? "ted" : create.source(), old.lolCount(), old.groanCount());
            return repo.update(id, updated);
        });
    }

    public boolean delete(String id) { return repo.delete(id); }

    public Optional<Joke> random(Optional<String> category) { return repo.random(category); }

    public Optional<Joke> bumpLol(String id) { return repo.bumpLol(id); }

    public Optional<Joke> bumpGroan(String id) { return repo.bumpGroan(id); }
}

File: nestjs/src/joke.service.ts
import {Injectable, NotFoundException} from '@nestjs/common';
import {InjectRepository} from '@nestjs/typeorm';
import {Repository} from 'typeorm';
import {Joke} from './joke.entity';
import {v4 as uuidv4} from 'uuid';

@Injectable()
export class JokeService {
  constructor(@InjectRepository(Joke) private repo: Repository<Joke>) {}

  async create(data: Partial<Joke>) {
    if (!data.setup || !data.punchline) throw new Error('setup and punchline required');
    if (!data.id) data.id = uuidv4();
    const j = this.repo.create(data as Joke);
    return this.repo.save(j);
  }

  async findById(id: string) {
    const j = await this.repo.findOne({where: {id}});
    if (!j) throw new NotFoundException();
    return j;
  }

  async update(id: string, body: Partial<Joke>) {
    const j = await this.findById(id);
    Object.assign(j, body);
    return this.repo.save(j);
  }

  async delete(id: string) {
    await this.repo.delete({id});
  }

  async bumpLol(id: string) {
    const j = await this.findById(id);
    j.lolCount = (j.lolCount || 0) + 1;
    return this.repo.save(j);
  }

  async bumpGroan(id: string) {
    const j = await this.findById(id);
    j.groanCount = (j.groanCount || 0) + 1;
    return this.repo.save(j);
  }

  async random() {
    const all = await this.repo.find();
    if (!all || all.length === 0) throw new NotFoundException();
    return all[Math.floor(Math.random() * all.length)];
  }

  async randomByCategory(category: string) {
    const list = await this.repo.find({where: {category}});
    if (!list || list.length === 0) throw new NotFoundException();
    return list[Math.floor(Math.random() * list.length)];
  }

  async bySource(source: string) {
    return this.repo.find({where: {source}});
  }
}
